fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/properties/DerivationsProperty.java

```diff
@@ -0,0 +1,789 @@
+            throw new RecordCoreException("unsupported plan operator");
+        }
+
+        @Nonnull
+        @Override
+        public Derivations visitIndexPlan(@Nonnull final RecordQueryIndexPlan indexPlan) {
+            final var matchCandidate = indexPlan.getMatchCandidate();
+            return visitPlanWithComparisons(indexPlan, matchCandidate.getQueriedRecordTypeNames());
+        }
+
+        @Nonnull
+        private Derivations visitPlanWithComparisons(@Nonnull final RecordQueryPlanWithComparisons planWithComparisons,
+                                                     @Nonnull final Iterable<String> recordTypeNames) {
+            final var comparisons = planWithComparisons.getComparisons();
+            final var comparisonValues = comparisons.stream()
+                    .flatMap(comparison -> comparison.getValues().stream())
+                    .collect(ImmutableList.toImmutableList());
+            final var resultValueFromPlan = planWithComparisons.getResultValue();
+            final var resultValue = new QueriedValue(resultValueFromPlan.getResultType(), recordTypeNames);
+
+            return new Derivations(ImmutableList.of(resultValue), comparisonValues);
+        }
+
+        @Nonnull
+        @Override
+        public Derivations visitFirstOrDefaultPlan(@Nonnull final RecordQueryFirstOrDefaultPlan firstOrDefaultPlan) {
+            final Quantifier rangesOver = Iterables.getOnlyElement(firstOrDefaultPlan.getQuantifiers());
+            final var childDerivations = derivationsFromSingleChild(firstOrDefaultPlan);
+            final var childResultValues = childDerivations.getResultValues();
+            final var onEmptyResultValue = firstOrDefaultPlan.getOnEmptyResultValue();
+            final var localValuesBuilder = ImmutableList.<Value>builder();
+            localValuesBuilder.addAll(childDerivations.getLocalValues());
+            for (final var childResultValue : childResultValues) {
+                final var resultsTranslationMap = TranslationMap.builder()
+                        .when(rangesOver.getAlias()).then(Quantifier.current(), ((sourceAlias, targetAlias, leafValue) -> childResultValue))
+                        .build();
+                localValuesBuilder.add(onEmptyResultValue.translateCorrelations(resultsTranslationMap));
+            }
+            return new Derivations(childDerivations.getResultValues(), localValuesBuilder.build());
+        }
+
+        @Nonnull
+        @SuppressWarnings("java:S135")
+        public Derivations visitInJoinPlan(@Nonnull final RecordQueryInJoinPlan inJoinPlan) {
+            final var outerAlias = inJoinPlan.getInAlias();
+            final var innerQuantifier = inJoinPlan.getInner();
+            final var innerDerivations = derivationsFromQuantifier(innerQuantifier);
+
+            //
+            // De-correlate inner against in-source. Since the in-source is not really a value, we can only
+            // fake it. We just need to remove the correlation.
```

Comment:
I am not sure I agree with this comment, removing the correlation here means replacing it with its `Value`, effectively "flattening" and denormalising all references (the same I used to mention to you few months ago), is this the case here or am I missing something?

Reply:
Can you please elaborate?

Reply:
I am not sure I am getting the sentiment of the comment, however, let me explain my own comment a bit better: Replacing a `QOV(q)` with its definition is not correct execution-wise. It only works in this case, because we are trying to establish a static derivative tree that can be traversed easily. The `Value` tree we produce here would only produce nonsense if executed. The `in-source` is a source for a correlation, (however it predates flat maps thus we have it in the first place). The `in-source` in the plan is not a `Value` thus I cannot replace a correlation referring to a binding coming from an in-source. Thus I am creating a fake value that I can use when decorrelating the inner of the plan.